### ISSUE TITLE
azurerm_sql_active_directory_administrator removed from azurerm_mssql_server on subsequent deployments

### ISSUE THREAD

**Issue by mikegouldthorp** [hashicorp/terraform-provider-azurerm#9946]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform = 0.12.29
azurerm = 2.41.0

### Affected Resource(s)

* `resource "azurerm_mssql_server"`
* `resource "azurerm_sql_active_directory_administrator"`

### Terraform Configuration Files

```hcl
resource "azurerm_mssql_server" "mssqlsrvr" {
  name                         = "mssqlsrvr"
  resource_group_name          = data.azurerm_resource_group.rg.name
  location                     = var.location
  version                      = "12.0"
  administrator_login          = azurerm_key_vault_secret.username.value
  administrator_login_password = azurerm_key_vault_secret.password.value
  minimum_tls_version          = "1.2"
}

resource "azurerm_sql_active_directory_administrator" "sql-admin" {
  server_name          = azurerm_mssql_server.mssqlsrvr.name
  resource_group_name  = azurerm_mssql_server.mssqlsrvr.resource_group_name
  login                = var.active_directory_admin_group
  object_id            = var.active_directory_admin_group_objid
  tenant_id            = "<redacted>"
}

```

### Expected Behavior

The Azure AD Administrator (group) is defined for the SQL server.  Subsequent deployments does not impact the Azure AD Administrator association.

### Actual Behavior

The first time the SQL server and the SQL administrator resources are provisioned the AD admin group is successfully assigned to the server. If the same resources are provisioned a second time the AD admin group is removed. Once the AD admin group is removed provisioning the resources a third time will add the AD admin group and so on.

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:

  ~ update in-place

Terraform will perform the following actions:

  # module.sql_db_single.azurerm_mssql_server.mssqlsrvr will be updated in-place
  ~ resource "azurerm_mssql_server" "mssqlsrvr" {
        administrator_login             = "<redacted>"
        administrator_login_password    = (sensitive value)
        connection_policy               = "Default"
        extended_auditing_policy        = []
        fully_qualified_domain_name     = "<redacted>"
        id                              = "<redacted>"
        location                        = "northcentralus"
        minimum_tls_version             = "1.2"
        name                            = "mssqlsrvr"
        public_network_access_enabled   = true
        resource_group_name             = "mssqlsrvr-rg"
        restorable_dropped_database_ids = []
        version                         = "12.0"

      - azuread_administrator {
          - login_username = "AD_SQLAdminGroup" -> null
          - object_id      = "<redacted>" -> null
          - tenant_id      = "***" -> null
```

### Steps to Reproduce

Provision the same `azurerm_mssql_server` resource and associated `azurerm_sql_active_directory_administrator` resource multiple times.

1. `terraform apply`

### References

* #0000

**Comment by mikegouldthorp** [comment]
Is there any additional information that I can provide to help this issue gain some traction?

**Comment by carluu** [comment]
I just stumbled across this looking for something else, but have you considered using the azuread_administrator {} block embedded within your mssql_server resource rather than having it separate? I'm referring to this: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/mssql_server#azuread_administrator

Right now it looks like you're trying to join two resource types (azurerm_mssql and azurerm_sql) so I'm guessing when you run the azurerm_mssql_server deployment on that second run, it doesn't know you specified an AD Admin in a different resource and is nulling those values out

**Comment by mikegouldthorp** [comment]
@carluu I took your advice and removed the azurerm_sql_active_directory_administrator resource and set the AD user\group using the azurerm_mssql_server azuread_administrator {} block.

What I found is that for **new** deployments of the azurerm_mssql_server resource the azuread_administrator {} block works. I can re-provision the resource multiple times and I don't encounter the issue where the AD admin group getting removed and reapplied on subsequent deployments.

Unfortunately there is still an issue if I attempt to remove the azurerm_sql_active_directory_administrator resource in favor of the azuread_administrator {} block for **existing** azurerm_mssql_server resources. 

I confirmed the azurerm_sql_active_directory_administrator resource is destroyed but the problem is that AD group assigned in the azuread_administrator {} block does NOT get applied. 

```
# module.mssql_db_single.azurerm_sql_active_directory_administrator.sql-admin will be destroyed
 - resource "azurerm_sql_active_directory_administrator" "sql-admin" {
   - id                  = "/<redacted>/administrators/activeDirectory" -> null
   - login               = "Database Admins" -> null
   - object_id           = "<redacted>" -> null
   - resource_group_name = "<redacted>" -> null
   - server_name         = "<redacted>" -> null
   - tenant_id           = "***" -> null
 }
```
I confirmed that the AD admin group is defined in the state file but when I look in the Azure portal it is not set.   Re-running the pipeline a second time states that there are no changes.

```{
    "module": "module.mssql_db_single",
    "mode": "managed",
    "type": "azurerm_mssql_server",
    "name": "mssqlsrvr",
    "provider": "provider.azurerm",
    "instances": [
      {
        "schema_version": 0,
        "attributes": {
          "administrator_login": "<redacted>",
          "administrator_login_password": "<redacted>",
          "azuread_administrator": [
            {
              "login_username": "Database Admins",
              "object_id": "<redacted>",
              "tenant_id": "<redacted>"
            }
          ],```

The following issue #10643 describes a similar issue with the azuread_administrator.

**Comment by RacerEXE** [comment]
I hit the second part of the bug you noted above. I have a azurerm_mssql_server resource that was provisioned without an AD administrator. I am trying to add an AD group as the administrator via the azuread_administrator {} block. The plan says it will update the azurerm_mssql_server, after the apply the statefile shows the azuread_administrator, but in the Azure portal it is missing.

Does anyone know any workarounds?

**Comment by RacerEXE** [comment]
**TL;DR**

@mikegouldthorp, I am not sure if this is your issue when trying to use the azuread_administrator. However, I found that the error described in the last two posts occurs if the azuread_administrator.login_username is the same as the administrator_login in the enclosing azurerm_mssql_server. **These values must be different!** If you try adding a azuread_administrator block to an already existing azurerm_mssql_server, the no error is given, however when deploying a new sql server where these values are the same gives a proper error.

**Details**

My terraform code originally tried to add the  azuread_administrator with administrator_login == azuread_administrator.login_username (sqladmin) to two existing databases (one in use2 and one in usw2). I tried deleting my usw2 databases, and recreating the database. The plan showed my usw2 database would get created with the proper azuread_administrator block. However, during the apply following error:

`Error: waiting for creation of SQL Server "db-musw2-sql" AAD admin (Resource Group "sql-musw2-data-rg"): Code="ServerAdministratorNameAlreadyExists" Message="The external admin cannot be set because the principal name sqladmin already exists in the 'master' database."

  on .terraform/modules/sql_test/main.tf line 125, in resource "azurerm_mssql_server" "this":
 125: resource "azurerm_mssql_server" "this" {
`

I updated my code to in the azuread_administrator such that azuread_administrator.login_username = group_sqladmin. As expected my plan showed that my usw2 database would get created with the azuread_administrator block and my use2 sql server would be updated in place:

'      ~ azuread_administrator {
          ~ login_username = "sqladmin" -> "group_sqladmin"
            # (1 unchanged attribute hidden)
        }
'
When I looked in the azure portal, much to my surprise, both my usw2 server (which was being recreated from scratch), and my use2 server (which just had the login_username changed) were setup correctly with AD controlled admin roles.

My guess is that if you reuse the user id in both places, and you are updating a database, not creating a new one, the provider swallows the error or the azure API fails to return the error.


In my plan I saw that the data

**Comment by timwebster9** [comment]
I have a similar issue which might be related.

I created a server with an AAD admin block like so, and that worked fine:

```
  azuread_administrator {
    login_username              = azuread_group.example.display_name
    object_id                   = azuread_group.example.object_id
  }
```

Then I noticed that with 2.86 we can set the `azuread_authentication_only` flag, so I added it and re-ran Terraform:

```
  azuread_administrator {
    login_username              = azuread_group.example.display_name
    object_id                   = azuread_group.example.object_id
    azuread_authentication_only = true
  }
```

Terraform correctly detected that a change was required:

```
Terraform will perform the following actions:

  # module.sql.azurerm_mssql_server.aad_admin will be updated in-place
  ~ resource "azurerm_mssql_server" "aad_admin" {
        id                              = "/subscriptions/2ca65474-3b7b-40f2-b242-0d2fba4bde6e/resourceGroups/policy-rg/providers/Microsoft.Sql/servers/timwaadadmin"
        name                            = "timwaadadmin"
        tags                            = {}
        # (11 unchanged attributes hidden)

      ~ azuread_administrator {
          ~ azuread_authentication_only = false -> true
            # (3 unchanged attributes hidden)
        }

        # (1 unchanged block hidden)
    }
```

However in the actual request to update, the entire 'administrators' object was missing from the request:

```
PUT /subscriptions/2ca65474-3b7b-40f2-b242-0d2fba4bde6e/resourceGroups/policy-rg/providers/Microsoft.Sql/servers/timwaadadmin?api-version=2021-02-01-preview HTTP/1.1
Host: management.azure.com
User-Agent: Go/go1.17.3 (amd64-darwin) go-autorest/v14.2.1 Azure-SDK-For-Go/v59.0.0 sql/v5.0 HashiCorp Terraform/1.0.5 (+https://www.terraform.io) Terraform Plugin SDK/2.8.0 terraform-provider-azurerm/2.86.0 pid-222c6c49-1b0a-5959-a213-6608f9eb8820
Content-Length: 200
Content-Type: application/json; charset=utf-8
X-Ms-Correlation-Request-Id: 69787f8c-c23c-0e1b-4f87-734b4269aa1b
Accept-Encoding: gzip

{"identity":{"type":"SystemAssigned"},"location":"uksouth","properties":{"administratorLogin":"missadministrator","minimalTlsVersion":"1.2","publicNetworkAccess":"Enabled","version":"12.0"},"tags":{}}: timestamp=2021-11-23T18:06:45.788Z
```

I guess the result being that it tried to remove the AAD administrator entirely, and failed because I have a policy in place to enforce the use of an AAD Admin.

**Comment by HwanSolo** [comment]
I encountered the same issue as @mikegouldthorp for existing `azurerm_mssql_server` resources. 

I tried to apply the removing of the `azurerm_sql_active_directory_administrator` resource and setting the `azuread_administrator` on the `azurerm_mssql_server` resource block at the same time which resulted in no actual AD group.

What worked for me was to first apply the removal of the `azurerm_sql_active_directory_administrator` resource, then apply the new configuration with the `azuread_administrator` block. 

My existing `azurerm_mssql_server` resources now have the correct AD group and no longer get removed and reapplied on subsequent deployments.

**Comment by matt-byrne** [comment]
I have just encountered the same issue as above - After migrating from the old standard of both `azurerm_sql_server` and `azurerm_sql_active_directory_administrator` resources to the new method using `azurerm_mssql_server` with an `azuread_administrator` block, my AD administrator was removed and not re-created.

Since I'm using Terraform Cloud and everything applies at the same time, I can't use the workaround noted above either and have resorted to recreating the AD administrator after everything else has migrated.

**Comment by rcskosir** [comment]
Thanks for opening this issue. This was a problem in the 2.x version of the provider which is no longer actively maintained.
If this is still an issue with the 4.x version of the provider please do let us know by opening a new issue, thanks!
